### PR TITLE
Update token-cache-serialization.md

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 1]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
-
-```CSharp
-using Microsoft.Identity.Web;
-using Microsoft.Identity.Client;
-using Microsoft.Extensions.DependencyInjection;
-```
+## [Confidential Clients using MSAL](#tab/msal)
 
-```CSharp
+.NET confidential clients are recommended to use [Microsoft.Identity.Web](../microsoft-identity-web/index.md), which is based on MSAL.NET, as higher level APIs are able to handle complex scenarios out of the box (guest users, conditional access evaluation, proof of posession tokens etc.). Apps needing to access Azure APIs are recommended to use [Azure SDK](https://azure.github.io/azure-sdk/), which also leverage MSAL internally.
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:172]:
```suggestion
.NET confidential clients are recommended to use [Microsoft.Identity.Web](../microsoft-identity-web/index.md), which is based on MSAL.NET, as higher level APIs are able to handle complex scenarios out-of-the-box (e.g., guest users, Conditional Access Evaluation, Proof-of-Possession tokens). Applications that need to access Azure APIs should use [Azure SDK](https://azure.github.io/azure-sdk/), which leverages MSAL internally.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 2]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
-     {
-          return await app.AcquireTokenSilent(scopes, account).ExecuteAsync();; 
-     } 
-     catch (MsalUiRequiredException)
-     {
-      	// cannot get a token silently, so redirect the user to be challenged 
-     }
-  }
-```
+Read on if you're using MSAL.NET directly.
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:174]:
```suggestion
The following material is relevant if you're using MSAL.NET directly.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 3]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
 
 ### Available caching technologies
 
-Instead of `app.AddInMemoryTokenCache();`, you can use different caching serialization technologies. For example, you can use no-serialization, in-memory, and distributed token cache storage provided by .NET.
+- memory cache without eviction
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:178]:
```suggestion
- Memory cache without eviction
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 4]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
+- memory cache with eviction
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:179]:
```suggestion
- Memory cache with eviction
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 5]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
+- combined distributed cache and memory cache
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:180]:
```suggestion
- Combined distributed and memory caches
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 6]
```diff
@@ -160,77 +160,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
 
 <a id="no-token-cache-serialization"></a>
-#### Token cache without serialization 
+#### Memory Cache without eviction
 
-Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building the application and don't add any serializer.
+Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building a Service to Service application - `AcquireTokenForClient` that does not target many (>100k) tenants.
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:185]:
```suggestion
Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building a service-to-service application - `AcquireTokenForClient` that does not target many (more than 100,000) tenants.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 7]
```diff
@@ -248,37 +200,43 @@ Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building the
...
 
-If you use this approach and have a large number of users or tenants, be sure to monitor the memory footprint. If the memory footprint becomes a problem, consider enabling token cache serialization, which might reduce the internal cache size. Currently, you can't use shared cache and cache serialization together.
+### Memory Cache with eviction
 
-#### In-memory token cache
+Reference the [Microsoft.Identity.Web.TokenCache](https://www.nuget.org/packages/Microsoft.Identity.Web.TokenCache) NuGet package to your project.
 
-In-memory token cache serialization is great in samples. It's also good in production applications if you only request app tokens (`AcquireTokenForClient`), provided that you don't mind if the token cache is lost when the web app is restarted. We don't recommend it in production if you request user tokens (`AcquireTokenByAuthorizationCode`, `AcquireTokenSilent`, `AcquireTokenOnBehalfOf`).
+The following code shows how to add an in-memory cache with eviction
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:207]:
```suggestion
The following code shows how to add an in-memory cache with eviction:
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 8]
```diff
@@ -416,9 +374,7 @@ cacheHelper.RegisterCache(pca.UserTokenCache);
 
 #### Plain-text fallback mode
 
-The cross-platform token cache allows you to store unencrypted tokens in plain text. This feature is intended for use in development environments for debugging purposes only. 
-
-You can use the plain-text fallback mode by using the following code pattern.
+The cross-platform token cache allows you to store unencrypted tokens in an ACL restricted plain text file. This is useful in case the encryption at rest fails, which ocassionally does due to operating system environmnental reasons. You can use the plain-text fallback mode by using the following code pattern.
```
Comment (localden) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:377]:
```suggestion
The cross-platform token cache allows you to store unencrypted tokens in an ACL-restricted plain-text file. This is useful in cases where encryption at rest fails, which ocassionally happens due to environmnent-related reasons. You can use the plain-text fallback mode by using the following code pattern:
```

**Review by localden — CHANGES_REQUESTED** [review]

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 9]
```diff
@@ -23,32 +23,33 @@ After Microsoft Authentication Library (MSAL) [acquires a token](/azure/active-d
...
--	If you build an SDK and want to write your own token cache serializer for confidential client applications, inherit from [Microsoft.Identity.Web.MsalAbstractTokenCacheProvider](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs) and override the [WriteCacheBytesAsync](/dotnet/api/microsoft.identity.web.tokencacheproviders.msalabstracttokencacheprovider.writecachebytesasync) and [ReadCacheBytesAsync](/dotnet/api/microsoft.identity.web.tokencacheproviders.msalabstracttokencacheprovider.readcachebytesasync) methods.
+- When writing confidential client applications ([web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview), [web APIs](/azure/active-directory/develop/scenario-web-api-call-api-overview), or [service-to-service or daemon apps](/azure/active-directory/develop/scenario-daemon-overview), use [Microsoft.Identity.Web](../microsoft-identity-web/index.md) as a higher-level API. It offers integration with ASP.NET Core, ASP.NET Classic, and works standalone as well.
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:28]:
```suggestion
- When writing new confidential client applications ([web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview), [web APIs](/azure/active-directory/develop/scenario-web-api-call-api-overview), or [service-to-service or daemon apps](/azure/active-directory/develop/scenario-daemon-overview), use [Microsoft.Identity.Web](../microsoft-identity-web/index.md) as a higher-level API. It offers integration with ASP.NET Core, ASP.NET Classic, and works standalone as well.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 10]
```diff
@@ -23,32 +23,33 @@ After Microsoft Authentication Library (MSAL) [acquires a token](/azure/active-d
...
+- Existing confidential client applications that leverage MSAL.NET directly can continue to do so. 
+- [Web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview) and [web APIs](/azure/active-directory/develop/scenario-web-api-call-api-overview) should use a [distributed token cache](token-cache-serialization.md?tabs=aspnet#distributed-caches) (e.g., Redis, SQL Server, Azure Cosmos DB) in conjunction with a contrained memory cache.
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:30]:
```suggestion
- [Web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview) and [web APIs](/azure/active-directory/develop/scenario-web-api-call-api-overview) should use a [distributed token cache](token-cache-serialization.md?tabs=aspnet#distributed-caches) (e.g., Redis, SQL Server, Azure Cosmos DB) in conjunction with a constrained memory cache. 
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 11]
```diff
@@ -23,32 +23,33 @@ After Microsoft Authentication Library (MSAL) [acquires a token](/azure/active-d
...
+- Existing confidential client applications that leverage MSAL.NET directly can continue to do so. 
+- [Web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview) and [web APIs](/azure/active-directory/develop/scenario-web-api-call-api-overview) should use a [distributed token cache](token-cache-serialization.md?tabs=aspnet#distributed-caches) (e.g., Redis, SQL Server, Azure Cosmos DB) in conjunction with a contrained memory cache. 
+- Encryption at rest can be optionally configured using [ASP.NET Core Data Protection](/aspnet/core/security/data-protection/introduction).
+- [Web apps](/azure/active-directory/develop/scenario-web-app-call-api-overview) may also rely on session cookies; however, this option is not recommended due to cookie size.
+- [Service-to-service and daemon apps](token-cache-serialization.md?tabs=aspnet#distributed-caches) may rely on memory caching only. If your app serves many tenants, configure an eviction policy.
+- Managed Identity tokens are cached in memory only.
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:34]:
```suggestion
- Managed identity tokens are cached in memory only.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 12]
```diff
@@ -23,32 +23,33 @@ After Microsoft Authentication Library (MSAL) [acquires a token](/azure/active-d
...
 
+### Cache size
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:36]: Should this be added to the Size Approximations section below? The information is mostly the same.

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 13]
```diff
@@ -23,32 +23,33 @@ After Microsoft Authentication Library (MSAL) [acquires a token](/azure/active-d
...
 
-## [ASP.NET Core web apps and web APIs](#tab/aspnetcore)
+When users log in, there will be a cache entry for each user, around 7KB in size. The size will be larger if you are calling several downstream APIs. For service-to-service authentication, there will be a cache entry for each tenant and downstream API, around 2KB in size. 
 
-The [Microsoft.Identity.Web.TokenCache](https://www.nuget.org/packages/Microsoft.Identity.Web.TokenCache) NuGet package provides token cache serialization within the [Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web) library. 
+## [Confidential Clients using Microsoft.Identity.Web](#tab/aspnetcore)
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:40]:
```suggestion
## [Confidential clients using Microsoft.Identity.Web](#tab/aspnetcore)
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 14]
```diff
@@ -160,77 +161,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
-The following code shows how to add an in-memory well-partitioned token cache to your app.
-
-```CSharp
-using Microsoft.Identity.Web;
-using Microsoft.Identity.Client;
-using Microsoft.Extensions.DependencyInjection;
-```
-
-```CSharp
+## [Confidential Clients using MSAL](#tab/msal)
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:171]:
```suggestion
## [Confidential clients using MSAL.NET](#tab/msal)
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 15]
```diff
@@ -160,77 +161,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
-public static async Task<AuthenticationResult> GetTokenAsync(string clientId, X509Certificate cert, string authority, string[] scopes)
- {
-     // Create the confidential client application
-     app= ConfidentialClientApplicationBuilder.Create(clientId)
-       // Alternatively to the certificate, you can use .WithClientSecret(clientSecret)
-       .WithCertificate(cert)
-       .WithLegacyCacheCompatibility(false)
-       .WithAuthority(authority)
-       .Build();
+.NET confidential clients are recommended to use [Microsoft.Identity.Web](../microsoft-identity-web/index.md), which is based on MSAL.NET, as higher level APIs are able to handle complex scenarios out-of-the-box (e.g., guest users, Conditional Access Evaluation, Proof-of-Possession tokens). Applications that need to access Azure APIs should use [Azure SDK](https://azure.github.io/azure-sdk/), which leverages MSAL internally.
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:173]:
```suggestion
.NET confidential clients are recommended to use [Microsoft.Identity.Web](../microsoft-identity-web/index.md), which is based on MSAL.NET, as higher level APIs are able to handle complex scenarios out-of-the-box (e.g., guest users, Continuous Access Evaluation, Proof-of-Possession tokens). Applications that need to access Azure APIs should use [Azure SDK](https://azure.github.io/azure-sdk/), which leverages MSAL internally.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 16]
```diff
@@ -160,77 +161,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
 
 <a id="no-token-cache-serialization"></a>
-#### Token cache without serialization 
+#### Memory Cache without eviction
 
-Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building the application and don't add any serializer.
+Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building a service-to-service application - `AcquireTokenForClient` that does not target many (more than 100,000) tenants.
 
 > [!IMPORTANT]
 > There is no way to control the size of the cache with this option. If you are building a website, a web API, or a multi-tenant S2S app, then use the `In-memory token cache` option.
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:189]:
```suggestion
> There is no way to control the size of the cache with this option. If you are building a website, a web API, or a multi-tenant service-to-service app, then see the `Memory cache with eviction` section.
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 17]
```diff
@@ -160,77 +161,29 @@ services.AddCosmosCache((CosmosCacheOptions cacheOptions) =>
...
 ### Available caching technologies
 
-Instead of `app.AddInMemoryTokenCache();`, you can use different caching serialization technologies. For example, you can use no-serialization, in-memory, and distributed token cache storage provided by .NET.
+- Memory cache without eviction
+- Memory cache with eviction
+- Combined distributed and memory caches
 
 <a id="no-token-cache-serialization"></a>
-#### Token cache without serialization 
+#### Memory Cache without eviction
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:184]:
```suggestion
#### Memory cache without eviction
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 18]
```diff
@@ -248,37 +201,43 @@ Use `.WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` when building the
 
 `WithCacheOptions(CacheOptions.EnableSharedCacheOptions)` makes the internal MSAL token cache shared between MSAL client application instances. Sharing a token cache is faster than using any token cache serialization, but the internal in-memory token cache doesn't have eviction policies. Existing tokens are refreshed in place, but fetching tokens for different users, tenants, and resources makes the cache grow accordingly. 
 
-If you use this approach and have a large number of users or tenants, be sure to monitor the memory footprint. If the memory footprint becomes a problem, consider enabling token cache serialization, which might reduce the internal cache size. Currently, you can't use shared cache and cache serialization together.
+### Memory Cache with eviction
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:204]:
```suggestion
### Memory cache with eviction
```

File: msal-dotnet-articles/how-to/token-cache-serialization.md [hunk 19]
```diff
@@ -416,9 +375,7 @@ cacheHelper.RegisterCache(pca.UserTokenCache);
 
 #### Plain-text fallback mode
 
-The cross-platform token cache allows you to store unencrypted tokens in plain text. This feature is intended for use in development environments for debugging purposes only. 
-
-You can use the plain-text fallback mode by using the following code pattern.
+The cross-platform token cache allows you to store unencrypted tokens in an ACL-restricted plain-text file. This is useful in cases where encryption at rest fails, which ocassionally happens due to environmnent-related reasons. You can use the plain-text fallback mode by using the following code pattern:
```
Comment (pmaytak) [inline msal-dotnet-articles/how-to/token-cache-serialization.md:378]: Update this text above:

> In desktop applications, we recommend that you use the cross-platform token cache. MSAL.NET provides the cross-platform token cache in a separate library named [Microsoft.Identity.Client.Extensions.MSAL](https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet).

to this:
> In desktop applications, we recommend that you use the cross-platform token cache. MSAL.NET provides the cross-platform token cache in a separate library named [Microsoft.Identity.Client.Extensions.MSAL](https://www.nuget.org/packages/Microsoft.Identity.Client.Extensions.Msal).

Also update the references to the archives cache extensions repo to main MSAL repo.

Comment (localden) [reply msal-dotnet-articles/how-to/token-cache-serialization.md:378]: Added a separate issue for this: #363

**Pull request merged**